Cleanup and update of Decks.java

hssm opened the pull request:
The only real functional changes are in the deck renaming method which fixes a few issues and does error handling differently. The rest of the changes either minor fixes for edge cases that probably can never get triggered and to make the code more "pythonic" for maintainability (e.g., lists and iterators vs arrays and for loops).

Fixes include:
- Add missing "bury" key in default conf (we were doing `optBoolean("bury", true)` so no real impact)
- Don't allow renaming decks as subdecks of dyndecks
- Don't strip allowable characters from new deck names (fixes #3854)
- Don't crash on parsing empty deck names like  `deck::::deck3` (use same split logic as python)
- Add string resources for defaults (decks and confs)

For the sake of completeness I also added the missing bits for drag-drop deck rearrangement but most likely we won't get to implement a UI for it before this code is rewritten.

hssm commented:
The content provider tests are failing. I'm pretty sure it's because there's a context usage in a static string. I'll probably have to put it inside the constructor instead.

timrae commented:
How about adding a setter method where you give libanki a reference to the application context or something?

timrae commented on AnkiDroid/src/main/java/com/ichi2/libanki/Storage.java:
```diff
@@ -322,7 +323,7 @@ private static void _setColVars(AnkiDb db) {
             JSONObject g = new JSONObject(Decks.defaultDeck);
             g.put("id", 1);
-            g.put("name", "Default");
+            g.put("name", Utils.getResString(R.string.default_deck_name));
```
Presumably this is the problem

hssm commented:
At first I removed the resource reference from the static `defaultConf` template which made the tests work locally but not on travis. I removed the one you referred to after that and now it seems to work on both. I suppose it's not really a big loss since the default deck name gets hidden or renamed pretty early on anyway.

This is separate from my PR but for some reason one of the content provider tests fails on my device:
`testQueryNextCard: Check that we actually received one card expected:<1> but was:<0>`. Not sure what is going on with that but thought I'd mention it.

timrae commented:
To be honest I'm not a big fan of the static `getResString()` method, as it adds a cross dependency between libanki and the Application instance. 

Referencing the static getApplicationContext() from `DeckTask` is OK, but I feel that libanki should be more decoupled (i.e. the ApplicationContext should be passed in rather than statically retrieved from the ApplicationContext). It may work now, but as the API is expanded in the future I think it might introduce headaches.

timrae commented:
> for some reason one of the content provider tests fails on my device

Does it work with a completely fresh collection?

hssm commented:
Ah, yes, that was it.

timrae commented:
Btw, the tests will mess with your collection (force a full sync) so in
general you'll want to sandbox them
On 22/11/2015 12:34 pm, "Houssam Salem" <notifications@github.com> wrote:

> Ah, yes, that was it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ankidroid/Anki-Android/pull/3861#issuecomment-158705070>
> .
>

hssm commented:
In the end I removed most of the string resources. It seems to be more trouble than it's worth. I thought about passing in a context to Storage.java (essentially the libanki entry point) but it's used all over the place for things like importing and exporting and it's more work than I'm willing to do at this point.

>Btw, the tests will mess with your collection (force a full sync) so in general you'll want to sandbox them

Yeah I'm aware. I don't use my tablet for proper anki so it's pretty much my test device.

timrae commented on AnkiDroid/src/main/java/com/ichi2/anki/exception/DeckRenameException.java:
```diff
@@ -0,0 +1,9 @@
+public class DeckRenameException extends Exception {
+
+    public DeckRenameException(String msg) {
+        super(msg);
```
How about extending this exception class to something like below so that you don't have to give the Context to libanki?

``` java
public static final int ALREADY_EXISTS = 0;
private int mErrorCode;
public DeckRenameException(int errorCode) {
    mErrorCode = errorCode;
    super(msg);
}

public String getErrorMessage(Resources res) {
    switch (mErrorCode) {
        //     return res.getString()
    }
}
```

timrae commented on AnkiDroid/src/main/java/com/ichi2/libanki/Storage.java:
```diff
@@ -21,6 +21,7 @@
 
 import com.ichi2.anki.AnkiDatabaseManager;
 import com.ichi2.anki.AnkiDb;
+import com.ichi2.anki.R;
```
I think you forgot to delete this import

hssm commented:
That was a really good idea with the exception message. I've gone ahead and made that change.

timrae commented:
Great work!
